### ISSUE TITLE
unreachable!() code is reachable

### ISSUE THREAD

**Issue by jeancf** [riscv-rust/gd32vf103xx-hal#41]
In [this match statement](https://github.com/riscv-rust/gd32vf103xx-hal/blob/8db7a3b0780ad79cc6c06ea3de4b51e23c34cbc1/src/spi.rs#L118):

```
        let br = match self.base_freq.0 / freq.into().0 {
            0 => unreachable!(), 
            1..=2 => 0b000,
            3..=5 => 0b001,
            6..=11 => 0b010,
            12..=23 => 0b011,
            24..=47 => 0b100,
            48..=95 => 0b101,
            96..=191 => 0b110,
            _ => 0b111,
        };
```

The `unreachable!()` macro is actually reachable if the user specifies a desired frequency (integer) that is higher than the base frequency (integer). In this case, the execution of the `unreachable!()` macro terminates the execution with a panic!

This does not have to be the case. If the match statement is modified like below, specifying a frequency higher than the base frequency will set the clock at the base frequency and not panic:

```
        let br = match self.base_freq.0 / freq.into().0 {
            0..=2 => 0b000,
            3..=5 => 0b001,
            6..=11 => 0b010,
            12..=23 => 0b011,
            24..=47 => 0b100,
            48..=95 => 0b101,
            96..=191 => 0b110,
            _ => 0b111,
        };
```

I can submit a PR if you agree.

**Comment by Disasm** [comment]
I think this is in invalid situation anyway, but I agree that `unreachable!()` is not correct here. Maybe it should be replaced with something like `panic!("SPI frequency is too high")`.

**Comment by dkhayes117** [comment]
Issue resolved with PR #42
